tests/reproducibility/test_deep_graph_infomax.py

```diff
@@ -0,0 +1,61 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from .fixtures import assert_reproducible
+from stellargraph.layer import DeepGraphInfomax, GCN, APPNP, GAT, PPNP
+from stellargraph.mapper import FullBatchNodeGenerator, CorruptedGenerator
+from ..test_utils.graphs import example_graph_random
+import tensorflow as tf
+import pytest
+import numpy as np
+
+
+def dgi(generator, gen, model_type):
+
+    tf.random.set_seed(1234)
+    np.random.seed(1234)
+
+    emb_dim = 4
+
+    base_model = model_type(
+        generator=generator, activations=["relu"], layer_sizes=[emb_dim]
+    )
+    infomax = DeepGraphInfomax(base_model)
+
+    model = tf.keras.Model(*infomax.build())
+    model.compile(loss=tf.nn.sigmoid_cross_entropy_with_logits, optimizer="Adam")
+    model.fit(gen)
+    return model
+
+
+@pytest.mark.parametrize("model_type", [GCN, APPNP, GAT])
+def test_dgi_sparse(model_type):
+    G = example_graph_random()
+    generator = FullBatchNodeGenerator(G, sparse=True)
+    corrupted_generator = CorruptedGenerator(generator)
+    gen = corrupted_generator.flow(G.nodes())
+
+    assert_reproducible(lambda: dgi(generator, gen, model_type), num_iter=3)
+
+
+@pytest.mark.parametrize("model_type", [GCN, APPNP, GAT, PPNP])
+def test_dgi_dense(model_type):
+    G = example_graph_random()
+    generator = FullBatchNodeGenerator(G, sparse=True)
```

Comment:
```suggestion
    generator = FullBatchNodeGenerator(G, sparse=False)
```

I think?

Alternatively, you could parameterise (using [skipping](http://doc.pytest.org/en/latest/skipping.html#skipping-test-functions)):

```python
@pytest.mark.parametrize("model_type", [GCN, APPNP, GAT, PPNP])
@pytest.mark.parametrize("sparse", [False, True])
def test_dgi(model_type, sparse):
    if sparse and model is PPNP:
        pytest.skip("PPNP doesn't support sparse=True")

    G = example_graph_random()
    generator = FullBatchNodeGenerator(G, sparse=sparse)
    ...
```

(We could do a similar thing to reduce the duplication in the main tests, too.)

Also, this test seems to be currently running with `sparse=True` and passing for `PPNP`; are we sure that it doesn't support sparseness?

Reply:
`PPNP()` won't throw and error if passed a generator with `Sparse=True` but the results would be meaningless (maybe it should crash though?). So I don't think its worth testing as this may fail in the future. fyi creating `FullBatchNodeGenerator(G, sparse=True, method="ppnp")` will throw an error

Reply:
> PPNP() won't throw and error if passed a generator with Sparse=True but the results would be meaningless (maybe it should crash though?).

If it gives nonsense results, and we can catch it easily, yeah, we should. Should be a separate issue/PR though.

> (We could do a similar thing to reduce the duplication in the main tests, too.)

Did you try this?

Reply:
I'll make a seperate issue/PR
> Did you try this?
and yep, I put in there - much nicer! didn't realize you got stack pytest.param like that

Reply:
> I'll make a seperate issue/PR

👍 

> and yep, I put in there - much nicer! didn't realize you got stack pytest.param like that

I saw that you put it into `tests/reproducibility/test_deep_graph_infomax.py`, that parenthetical and my quote was about the main tests in `tests/layer/test_deep_graph_infomax.py`.

Reply:
oh forgot to do it for `layer/test_deep_graph_infomax.py`  my bad, I've added it in now